Sources/SourceKitLSP/Swift/CodeActions/ApplyDeMorganLaw.swift

```diff
@@ -0,0 +1,730 @@
+        return nil
+      }
+
+      let negatedExpr = ExprSyntax(
+        infixExpr.with(
+          \.operator,
+          ExprSyntax(biOperator.with(\.operator.tokenKind, .binaryOperator(comparisonOperator.negated.rawValue)))
+        )
+      )
+
+      return (self.map(negatedExpr: negatedExpr), .comparison)
+    }
+
+    /// If the expression of interest is a ternary expression, returns its negated expression and the type of change as
+    /// either `Change.negation` or `Change.ternary`.
+    ///
+    /// A ternary expression is negated to a negation expression if both of its sub-expressions will become negation
+    /// expressions upon negation, e.g.
+    /// ```swift
+    /// // prefer
+    /// (a ? b : c) /* negated to */ !(a ? b : c)
+    /// // rather than
+    /// (a ? b : c) /* negated to */ (a ? !b : !c)
+    /// ```
+    ///
+    /// Otherwise the negation will propagate to its sub-expressions, e.g.
+    /// ```swift
+    /// (a ? !b : c) /* negated to */ (a ? b : !c)
+    /// (a ? !b : !c) /* negated to */ (a ? b : c)
+    /// ```
+    func negatedTernary(exprType: ExprType) -> Negated? {
+      if let ternaryExpr = self.as(TernaryExprSyntax.self) {
+        switch NegatedTernary(ternaryExpr, exprType: exprType) {
+        case .negation(let negation):
+          self.map(negation: negation)
+        case .ternary(let ternary):
+          (self.map(negatedExpr: ternary), .ternary)
+        }
+      } else {
+        nil
+      }
+    }
+
+    /// The expression of interest for manipulation.
+    var exprOfInterest: ExprSyntax {
+      switch self {
+      case .single(_, let exprOfInterest), .other(let exprOfInterest):
+        exprOfInterest
+      }
+    }
+
+    /// Downcasts the expression of interest as the given `type`.
+    func `as`<E: ExprSyntaxProtocol>(_ type: E.Type) -> E? {
+      self.exprOfInterest.as(E.self)
+    }
+
+    /// Substitutes the expression of interest with the negated expression, prepend the negation prefix, and returns
+    /// the result.
+    ///
+    /// To avoid creating extraneous parentheses, we only add a pair of parantheses to the negated expression when,
+    ///  1. The negated expression is a composite expression, and
+    ///  2. The original expression is not a single.
+    ///
+    /// - Note: A composite expression is one of `AsExprSyntax`, `AwaitExprSyntax`, `InfixOperatorExprSyntax`,
+    /// `IsExprSyntax`, or `TryExprSyntax`.
+    ///
+    /// An example of mapping is as follows,
+    /// ```swift
+    /// // original expression
+    /// ((!a || b is String))
+    /// // expression of interest
+    /// !a || b is String
+    /// !a /* negated to and mapped to */ a
+    /// b is String /* negated to */ Negation(b is String) /* mapped to */ !(b is String)
+    /// // negated expression of interest
+    /// a && !(b is String)
+    /// // mapped to
+    /// !((a && !(b is String)))
+    /// ```
+    func map(negation: Negation) -> Negated {
```

Comment:
`map` seems like a pretty ambiguous term here but I’m struggling to find anything more meaningful. My feeling is also that the `negation` parameter and `self` duplicate some information. Just a thought: Could `Negation` just store the underlying `PrefixOperatorExprSyntax` that it was created from and then offer `exprType` and `expr` as computed properties (or stored properties if that’s better for performance)?

Reply:
I stored the negated expr of `PrefixOperatorExprSyntax` directly into `Negation` because it's what all the intermediate data types do. It's okay to turn `exprType` and `expr` into computed properties.

Reply:
> `map` seems like a pretty ambiguous term here but I’m struggling to find anything more meaningful.

I had functors in mind when choosing `map` as the name, regrettably I also struggled with finding a better name.